src/map.cpp

```diff
@@ -2573,12 +2573,16 @@ ServerMapSector * ServerMap::createSector(v2s16 p2d)
 	/*
 		Do not create over-limit
 	*/
-	const static u16 map_gen_limit = MYMIN(MAX_MAP_GENERATION_LIMIT,
+	const static u16 max_map_gen_limit = MYMIN(MAX_MAP_GENERATION_LIMIT,
 		g_settings->getU16("map_generation_limit"));
-	if(p2d.X < -map_gen_limit / MAP_BLOCKSIZE
-			|| p2d.X >  map_gen_limit / MAP_BLOCKSIZE
-			|| p2d.Y < -map_gen_limit / MAP_BLOCKSIZE
-			|| p2d.Y >  map_gen_limit / MAP_BLOCKSIZE)
+	const static v3s16 map_gen_volume = g_settings->getV3S16("map_generation_volume");
+
+	const static u16 map_gen_limit_x = MYMIN(max_map_gen_limit, map_gen_volume.X);
+	const static u16 map_gen_limit_z = MYMIN(max_map_gen_limit, map_gen_volume.Z);
+	if(p2d.X < -map_gen_limit_x / MAP_BLOCKSIZE
+	|| p2d.X >  map_gen_limit_x / MAP_BLOCKSIZE
```

Comment:
We need double indentation here, due to code style: http://dev.minetest.net/Code_style_guidelines